drivers/adc/adc_demo/iio_adc_demo.h

```diff
@@ -42,10 +42,45 @@
 
 #include <stdlib.h>
 #include "adc_demo.h"
+#include "iio.h"
 
-struct iio_device adc_demo_iio_descriptor = {
-	.debug_reg_read = (int32_t (*)()) adc_demo_reg_read,
-	.debug_reg_write = (int32_t (*)()) adc_demo_reg_write
+#define MAX_NR_CHANNELS 10
+#define MAX_NR_ATTRIBUTES 5
+#define TEST_CHANNEL_NO 2
+#define MAX_SAMPLES_PER_CHANNEL 800
+
+extern struct iio_channel iio_adc_channels[MAX_NR_CHANNELS];
+
+extern struct iio_attribute adc_channel_attributes[MAX_NR_ATTRIBUTES];
+
+extern struct iio_attribute iio_adc_global_attributes[MAX_NR_ATTRIBUTES];
+
+enum iio_adc_demo_attributes {
+	ADC_CHANNEL_ATTR,
+	ADC_GLOBAL_ATTR,
 };
 
+#define ADC_DEMO_ATTR(_name, _priv) {\
+	.name = _name,\
+	.priv = _priv,\
+	.show = get_adc_demo_attr,\
+	.store = set_adc_demo_attr \
+}
+
+#define ADC_DEMO_DEV(_numch) {\
+	.num_ch = _numch, \
+	.channels = iio_adc_channels, \
+	.attributes = iio_adc_global_attributes,	\
+	.debug_attributes = NULL,	\
+	.buffer_attributes = NULL,	\
+	.prepare_transfer = update_adc_channels,	\
+	.end_transfer = close_adc_channels,	\
+	.read_dev = (int32_t (*)())adc_read_samples,	\
+	.debug_reg_read = (int32_t (*)()) adc_demo_reg_read,	\
+	.debug_reg_write = (int32_t (*)()) adc_demo_reg_write	\
+}
+
+
+int32_t init_adc_channels(struct adc_demo_desc* dev, uint32_t mask);
```

Comment:
I would initialize as here https://github.com/analogdevicesinc/no-OS/blob/60c83ad3b11fe54de086746dce6607301399316b/iio/iio_adxrs290/iio_adxrs290.h#L144 for examples.
iio_adc_demo and iio_dac_demo should serve as example for new drivers and be as simple as possible. 
Therefore I think is better to use a fixed number of channels because the chips we are creating drivers for have a fixed number too.

Reply:
Having the option to choose a different number of channels would help us testing the entire infrastructure much easier - we want to make sure everything works even if for example a device has 1 channel or has 16 channels.